### PR TITLE
Fix/api duplicate xml root 3149

### PR DESCRIPTION
**Pull request by machisuji**
Fixes duplicate xml root elements [1] caused by both unexpected rabl behaviour (possibly a bug) and incorrect rabl usage.

[1] https://www.openproject.org/work_packages/3149

### CODE REVIEW

**Comment by ulferts** [comment]
Gemfile/Gemfile.lock seem to be out of sync.

**Comment by hschink** [comment]
Rebase

File: Gemfile [hunk 1]
```diff
@@ -64,7 +64,7 @@ gem 'therubyracer'
 gem "date_validator"
 
 # replacing rsb with rabl
-gem 'rabl'
+gem 'rabl', '0.9.3'
```
Comment (tessi) [inline Gemfile]: I'd like to have a comment why we rely on that specific version (and maybe a link to  config/initializers/rabl_hack.rb for explaination)

File: config/initializers/rabl_hack.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      end
+      to_hash_without_hack(options)
+    end
+
+    alias_method :to_hash_without_hack, :to_hash
+    alias_method :to_hash, :to_hash_with_hack
+  end
+end
+
+module Rabl
```
Comment (tessi) [inline config/initializers/rabl_hack.rb]: It's not a deal breaker, but you don't need to open the `Rabl` module twice :)

File: config/initializers/rabl_init.rb [hunk 3]
```diff
@@ -1,23 +1,24 @@
 Rabl.configure do |config|
+  config.json_engine = ::Oj # Class with #dump class method (defaults JSON)
+  config.include_json_root = true
```
Comment (tessi) [inline config/initializers/rabl_init.rb:3]: Why is json_root `true` when it was `false` before? I thought this change is about XML things

**Comment by tessi** [comment]
Can you please resolve any merge conflicts, so that this PR is mergable again?

File: config/initializers/rabl_init.rb (hunk 3 above)
Comment (machisuji) [inline config/initializers/rabl_init.rb:3]: It's because of 'config.include_child_root' which does not work correctly. I.e. when you set that you would expect no child root in collections. Which didn't work for the #collection call in RABL.
This caused what should look like this:

```
GET .../project.json
=>
{
  "projects": [
    {
      "id": 42,
      /* ... */
    },
  /* ... */
  ]
}
```

to look like this:

```
GET .../project.json
=>
{
  "projects": [
    "project": {
      "id": 42,
      /* ... */
    },
  /* ... */
  ]
}
```

The person that applied RABL in OpenProject originally worked around that bug through misusing RABL. They disabled 'include root' for both XML and JSON since only then would collections not have root elements either. But this broke the #show views which should look like this:

```
GET .../project/42.json
=>
{
  "project": {
    "id": 42,
    /* ... */
  }
}
```

By disabling the root in general, however, those were missing roots as well as the collection items. While it is desired with collections it is not with single objects shown. Hence then the show views had explicit #node calls where there shouldn't be any necessary. This fixed the JSON views but broke the XML ones (=> duplicate root). Also it made the code ugly.

So what I did is fix RABL to work as intended and then the views to use RABL as intended, too.

To summarise:
- single (root) objects should have a root node, collection items should not
- hence the following RABL options should be used:
  - include_xml_root = false, include_json_root = true
  - include_child_root = false

All this terminology is confusing for XML because in XML root and collection items _always_ need a root node by definition. So even with include_xml_root being false there will obviously be root nodes in the XML views. Set it to true and you'll get duplicate roots again.

File: config/initializers/rabl_hack.rb (hunk 2 above)
Comment (machisuji) [inline config/initializers/rabl_hack.rb]: Going to fix it.

File: Gemfile (hunk 1 above)
Comment (machisuji) [inline Gemfile]: Aye, aye.

File: config/initializers/rabl_init.rb (hunk 3 above)
Comment (tessi) [inline config/initializers/rabl_init.rb:3]: And why is `config.include_xml_root  = false` then when it should be `true`? (and thanks for the explaination! :))

Comment (machisuji) [inline config/initializers/rabl_init.rb:3]: Because rabl is silly.
